library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedJoinOperator.java

```diff
@@ -0,0 +1,206 @@
+  {
+    if (timestampExtractor2 == null) {
+      if (tuple instanceof Tuple.TimestampedTuple) {
+        return ((Tuple.TimestampedTuple)tuple).getTimestamp();
+      } else {
+        return 0;
+      }
+    } else {
+      return timestampExtractor2.apply(tuple.getValue());
+    }
+  }
+
+  public void dropTuple2(Tuple<InputT2> input)
+  {
+    // do nothing
+    LOG.debug("Dropping late tuple {}", input);
+  }
+
+  public Tuple.WindowedTuple<InputT2> getWindowedValue2(Tuple<InputT2> input)
+  {
+    // windowOption can be null here, but the two upstream windowed operators must have same window option.
+    // TODO: Perform window option check during validation. 
+    if (windowOption == null && input instanceof Tuple.WindowedTuple) {
+      // inherit the windows from upstream
+      for (Window window : ((Tuple.WindowedTuple<InputT2>)input).getWindows()) {
+        if (!windowStateMap.containsWindow(window)) {
+          windowStateMap.put(window, new WindowState());
+        }
+      }
+      return (Tuple.WindowedTuple<InputT2>)input;
+    } else {
+      return new Tuple.WindowedTuple<>(assignWindows2(input), extractTimestamp2(input), input.getValue());
+    }
+  }
+
+  private Collection<? extends Window> assignWindows2(Tuple<InputT2> inputTuple)
+  {
+    if (windowOption instanceof WindowOption.GlobalWindow) {
+      return GLOBAL_WINDOW_SINGLETON_SET;
+    } else {
+      long timestamp = extractTimestamp2(inputTuple);
+      if (windowOption instanceof WindowOption.TimeWindows) {
+        Collection<? extends Window> windows = getTimeWindowsForTimestamp(timestamp);
+        for (Window window : windows) {
+          if (!windowStateMap.containsWindow(window)) {
+            windowStateMap.put(window, new WindowState());
+          }
+        }
+        return windows;
+      } else if (windowOption instanceof WindowOption.SessionWindows) {
+        return assignSessionWindows2(timestamp, inputTuple);
+      } else {
+        throw new IllegalStateException("Unsupported Window Option: " + windowOption.getClass());
+      }
+    }
+  }
+
+  public void setTimestampExtractor2(Function<InputT2, Long> timestampExtractor2)
+  {
+    this.timestampExtractor2 = timestampExtractor2;
+  }
+
+  protected Collection<Window.SessionWindow> assignSessionWindows2(long timestamp, Tuple<InputT2> inputTuple)
+  {
+    throw new UnsupportedOperationException("Session window require keyed tuples");
+  }
+
+
+  @Override
+  public void processWatermark(ControlTuple.Watermark watermark)
+  {
+    latestWatermark1 = watermark.getTimestamp();
+    if (this.watermarkTimestamp < 0) {
+      this.watermarkTimestamp = watermark.getTimestamp();
+    } else {
+      // Select the smallest timestamp of the latest watermarks as the watermark of the operator.
+      if (Math.min(latestWatermark1, latestWatermark2) < this.watermarkTimestamp) {
+        this.watermarkTimestamp = Math.min(latestWatermark1, latestWatermark2);
```

Comment:
If the previous watermarkTimestamp (in previous Apex window) was not changed, this would still trigger processWatermark at endWindow, which is not the right behavior.